mesonbuild/wrap/wrap.py

```diff
@@ -166,107 +176,73 @@ def resolve_git_submodule(self, dirname):
             return False
         # Submodule has not been added, add it
         if out.startswith(b'+'):
-            mlog.warning('git submodule {} might be out of date'.format(dirname))
+            mlog.warning('git submodule might be out of date')
             return True
         elif out.startswith(b'U'):
-            raise RuntimeError('submodule {} has merge conflicts'.format(dirname))
+            raise WrapException('git submodule has merge conflicts')
         # Submodule exists, but is deinitialized or wasn't initialized
         elif out.startswith(b'-'):
             if subprocess.call(['git', '-C', self.subdir_root, 'submodule', 'update', '--init', dirname]) == 0:
                 return True
-            raise RuntimeError('Failed to git submodule init {!r}'.format(dirname))
+            raise WrapException('git submodule failed to init')
         # Submodule looks fine, but maybe it wasn't populated properly. Do a checkout.
         elif out.startswith(b' '):
             subprocess.call(['git', 'checkout', '.'], cwd=dirname)
             # Even if checkout failed, try building it anyway and let the user
             # handle any problems manually.
             return True
+        elif out == b'':
+            # It is not a submodule, just a folder that exists in the main repository.
+            return False
         m = 'Unknown git submodule output: {!r}'
-        raise RuntimeError(m.format(out))
+        raise WrapException(m.format(out))
+
+    def get_file(self, p):
+        path = self.get_file_internal(p, 'source')
+        target_dir = os.path.join(self.subdir_root, p.get('directory'))
```

Comment:
Should this (and similar ones below) use `self.directory`?

Reply:
I removed all those p.get('directory') and replaced by self.dirname. That has a side effect that the 'directory' key is now optional in the wrap file, and will default to packagename. IMHO that was uselessly redundant info anyway.

Reply:
It was really hard to tell if it was supposed to be optional or not before. The wrap file format really needs real documentation. But that of course is a different project^^